plumbing/object/commitgraph/commitnode_walker_helper.go

```diff
@@ -0,0 +1,156 @@
+package commitgraph
+
+import (
+	"math"
+
+	"github.com/go-git/go-git/v5/plumbing"
+
+	"github.com/emirpasic/gods/trees/binaryheap"
+)
+
+// commitNodeStackable represents a common interface between heaps and stacks
+type commitNodeStackable interface {
+	Push(c CommitNode)
+	Pop() (CommitNode, bool)
+	Peek() (CommitNode, bool)
+	Size() int
+}
+
+// commitNodeLifo is a stack implementation using an underlying slice
+type commitNodeLifo struct {
+	l []CommitNode
+}
+
+// Push pushes a new CommitNode to the stack
+func (l *commitNodeLifo) Push(c CommitNode) {
+	l.l = append(l.l, c)
+}
+
+// Pop pops the most recently added CommitNode from the stack
+func (l *commitNodeLifo) Pop() (CommitNode, bool) {
+	if len(l.l) == 0 {
+		return nil, false
+	}
+	c := l.l[len(l.l)-1]
+	l.l = l.l[:len(l.l)-1]
```

Comment:
I suspect this could cause a "slice memory leak".

Reply:
Yes, The slice that is used will not reduce in capacity but the slice only exists for the time that the `commitNodeLifo` exists.

This is only used in `NewCommitNodeIterTopoOrder` as the stack for the nodes to that are ready to be printed. Therefore the slice should not get too large and would tend to fluctuate between 1 & 2 nodes up to the max number of active branches for a commit - which would tend to be a similar across the history.

We could make a more complex Lifo, one that say, has a permanent minimum capacity of 8 and if the length falls below a quarter of the current capacity it would reduce to half capacity if preferred

Reply:
Let's keep as is for simplicity - we can revisit it later if worst comes to worst.